q2_annotate/busco/utils.py

```diff
@@ -323,3 +402,103 @@ def _process_busco_results(results, sample_id, mag_id, file_name, additional_met
     }
 
     return results
+
+
+def _get_mag_lengths(bins: Union[MultiMAGSequencesDirFmt, MAGSequencesDirFmt]):
+    lengths = {}
+    if isinstance(bins, MultiMAGSequencesDirFmt):
+        for sample, mags in bins.sample_dict().items():
+            for mag_id, mag_fp in mags.items():
+                seq = skbio.io.read(mag_fp, format="fasta")
+                lengths[mag_id] = sum([len(s) for s in seq])
+        return pd.Series(lengths, name="length")
+    else:
+        for mag_id, mag_fp in bins.feature_dict().items():
+            seq = skbio.io.read(mag_fp, format="fasta")
+            lengths[mag_id] = sum([len(s) for s in seq])
+        return pd.Series(lengths, name="length")
+
+
+def _filter_unbinned_for_partition(unbinned_contigs, mag_partition, _filter_contigs):
+    """
+    Filters the unbinned contigs to match the sample IDs in a MAG partition.
+
+    Args:
+        unbinned_contigs (ContigSequencesDirFmt): The full unbinned contigs.
+        mag_partition (MultiMAGSequencesDirFmt): One partition of MAGs.
+        _filter_contigs (Action): QIIME 2 action to filter contigs.
+
+    Returns:
+        ContigSequencesDirFmt: Filtered unbinned contigs matching the partition samples.
+    """
+    sample_ids = list(mag_partition.view(MultiMAGSequencesDirFmt).sample_dict().keys())
+    metadata = Metadata(pd.DataFrame(index=pd.Index(sample_ids, name="ID")))
+    id_list = ", ".join([f"'{sid}'" for sid in sample_ids])
```

Comment:
I know we've decided to do it this way but I actually fixed this "bug" in q2-assembly so it should now be possible to pass the metadata without the "where" statement. Could you please simplify this here too? You may need to install the latest q2-assembly using pip to get my change into your conda env.